File: cute_framework/src/ch/hsr/ifs/test/framework/ui/TestRunnerViewPart.java
/*******************************************************************************
 * Copyright (c) 2007 Institute for Software, HSR Hochschule für Technik  
 * Rapperswil, University of applied sciences
 * All rights reserved. This program and the accompanying materials 
 * are made available under the terms of the Eclipse Public License v1.0 
 * which accompanies this distribution, and is available at 
 * http://www.eclipse.org/legal/epl-v10.html  
 * 
 * Contributors: 
 * Emanuel Graf - initial API and implementation 
 ******************************************************************************/
package ch.hsr.ifs.test.framework.ui;

import org.eclipse.core.runtime.IProgressMonitor;
import org.eclipse.core.runtime.IStatus;
import org.eclipse.core.runtime.Status;
import org.eclipse.debug.core.DebugException;
import org.eclipse.debug.core.ILaunchConfiguration;
import org.eclipse.debug.core.model.IProcess;
import org.eclipse.debug.ui.DebugUITools;
import org.eclipse.jface.action.Action;
import org.eclipse.jface.action.IToolBarManager;
import org.eclipse.jface.action.Separator;
import org.eclipse.swt.SWT;
import org.eclipse.swt.events.ControlEvent;
import org.eclipse.swt.events.ControlListener;
import org.eclipse.swt.graphics.Point;
import org.eclipse.swt.layout.GridData;
import org.eclipse.swt.layout.GridLayout;
import org.eclipse.swt.widgets.Composite;
import org.eclipse.ui.IActionBars;
import org.eclipse.ui.part.ViewPart;
import org.eclipse.ui.progress.UIJob;

import ch.hsr.ifs.test.framework.Messages;
import ch.hsr.ifs.test.framework.TestFrameworkPlugin;
import ch.hsr.ifs.test.framework.ImageProvider;
import ch.hsr.ifs.test.framework.model.ISessionListener;
import ch.hsr.ifs.test.framework.model.TestSession;

public class TestRunnerViewPart extends ViewPart implements ISessionListener {
	
	private enum Orientation{horizontal, vertical}; 

	public static final String ID = "ch.hsr.ifs.cutelauncher.ui.TestRunnerViewPart"; //$NON-NLS-1$

	private Composite top = null;

	private Composite TopPanel = null;
	
	protected boolean autoScroll = true;

	private CounterPanel counterPanel = null;

	private CuteProgressBar cuteProgressBar = null;

	private TestViewer testViewer = null;
	
	private Composite parent;
	
	private Orientation currentOrientation = Orientation.horizontal;

	private ScrollLockAction scrollLockAction;
	private FailuresOnlyFilterAction failureOnlyAction;
	private Action showNextFailureAction;
	private Action showPreviousFailureAction;
	private RerunLastTestAction rerunLastTestAction;

	private TestSession session;

	private StopAction stopAction;
	private static Messages msg = TestFrameworkPlugin.getMessages();
	

	public TestRunnerViewPart() {
		super();
		TestFrameworkPlugin.getModel().addListener(this);		
	}

	@Override
	public void createPartControl(Composite parent) {
		this.parent = parent;
		addResizeListener(parent);
		GridLayout gridLayout = new GridLayout();
		gridLayout.numColumns = 1;
		gridLayout.marginWidth = 0;
		gridLayout.horizontalSpacing = 0;
		gridLayout.marginWidth = 0;
		gridLayout.marginHeight = 0;
		GridData gdata = new GridData();
		gdata.grabExcessHorizontalSpace = true;
		top = new Composite(parent, SWT.NONE);
		top.setLayout(gridLayout);
		top.setLayoutData(gdata);
		createTopPanel();
		createTestViewer();
		configureToolbar();
		setPartName(msg.getString("TestRunnerViewPart.Name")); //$NON-NLS-1$
		setTitleImage(TestFrameworkPlugin.getImageProvider().getImage(ImageProvider.APP_LOGO).createImage());
	}
	
	private void addResizeListener(Composite parent) {
		parent.addControlListener(new ControlListener() {
			public void controlMoved(ControlEvent e) {
			}
			public void controlResized(ControlEvent e) {
				computeOrientation();
			}
		});
	}
	
	public boolean isCreated() {
		return counterPanel != null;
	}
	
	private void configureToolbar() {
		IActionBars actionBars= getViewSite().getActionBars();
		IToolBarManager toolBar= actionBars.getToolBarManager();
		
		scrollLockAction= new ScrollLockAction(this);
		scrollLockAction.setChecked(!autoScroll);
		
		failureOnlyAction = new FailuresOnlyFilterAction();
		failureOnlyAction.setChecked(false);
		
		showNextFailureAction = new ShowNextFailureAction(this);
		showNextFailureAction.setEnabled(false);
		showPreviousFailureAction = new ShowPreviousFailureAction(this);
		showPreviousFailureAction.setEnabled(false);
		
		rerunLastTestAction = new RerunLastTestAction();
		rerunLastTestAction.setEnabled(false);
		
		stopAction = new StopAction();
		stopAction.setEnabled(false);
		
		toolBar.add(showNextFailureAction);
		toolBar.add(showPreviousFailureAction);
		toolBar.add(failureOnlyAction);
		toolBar.add(scrollLockAction);
		toolBar.add(new Separator());
		toolBar.add(rerunLastTestAction);
		toolBar.add(stopAction);
	}

	/**
	 * This method initializes TopPSanel	
	 *
	 */
	private void createTopPanel() {
		GridLayout gridLayout1 = new GridLayout();
		gridLayout1.numColumns = 2;
		GridData gridData = new GridData();
		gridData.grabExcessHorizontalSpace = true;
		gridData.verticalAlignment = org.eclipse.swt.layout.GridData.FILL;
		gridData.horizontalAlignment = org.eclipse.swt.layout.GridData.FILL;
		TopPanel = new Composite(top, SWT.NONE);
		createCounterPanel();
		TopPanel.setLayout(gridLayout1);
		TopPanel.setLayoutData(gridData);
		createCuteProgressBar();
	}

	/**
	 * This method initializes counterPanel	
	 *
	 */
	private void createCounterPanel() {
		GridData gridData1 = new org.eclipse.swt.layout.GridData();
		gridData1.grabExcessHorizontalSpace = true;
		gridData1.horizontalAlignment = org.eclipse.swt.layout.GridData.FILL;
		counterPanel = new CounterPanel(TopPanel, SWT.NONE);
		counterPanel.setLayoutData(gridData1);
	}

	/**
	 * This method initializes cuteProgressBar	
	 *
	 */
	private void createCuteProgressBar() {
		GridData gridData2 = new GridData();
		gridData2.grabExcessHorizontalSpace = true;
		gridData2.horizontalIndent = 35;
		gridData2.horizontalAlignment = org.eclipse.swt.layout.GridData.FILL;
		cuteProgressBar = new CuteProgressBar(TopPanel);
		cuteProgressBar.setLayoutData(gridData2);
	}



	/**
	 * This method initializes testViewer	
	 *
	 */
	private void createTestViewer() {
		GridData gridData3 = new org.eclipse.swt.layout.GridData();
		gridData3.grabExcessHorizontalSpace = true;
		gridData3.horizontalAlignment = org.eclipse.swt.layout.GridData.FILL;
		gridData3.verticalAlignment = org.eclipse.swt.layout.GridData.FILL;
		gridData3.grabExcessVerticalSpace = true;
		testViewer = new TestViewer(top, SWT.NONE, this);
		testViewer.setLayoutData(gridData3);
	}
	
	private void computeOrientation() {
			Point size= parent.getSize();
			if (size.x != 0 && size.y != 0) {
				if (size.x > size.y) 
					setOrientation(Orientation.horizontal);
				else 
					setOrientation(Orientation.vertical);
			}
	}

	private void setOrientation(Orientation orientation) {
		testViewer.setOrientation(orientation == Orientation.horizontal);
		currentOrientation = orientation;
		GridLayout layout= (GridLayout) TopPanel.getLayout();
		setCounterColumns(layout); 
		parent.layout();
	}

	private void setCounterColumns(GridLayout layout) {
		if (currentOrientation == Orientation.horizontal)
			layout.numColumns= 2; 
		else
			layout.numColumns= 1;
	}

	@Override
	public void setFocus() {
	}

	public boolean isAutoScroll() {
		return autoScroll;
	}

	public void setAutoScroll(boolean autoScroll) {
		this.autoScroll = autoScroll;
	}
	
	private final class SessionFinishedUIJob extends UIJob {
		private SessionFinishedUIJob(String name) {
			super(name);
		}

		@Override
		public IStatus runInUIThread(IProgressMonitor monitor) {
			rerunLastTestAction.setEnabled(true);
			stopAction.setEnabled(false);
			if(TestRunnerViewPart.this.session.hasErrorOrFailure()) {
				showNextFailureAction.setEnabled(true);
				showPreviousFailureAction.setEnabled(true);
				if(isAutoScroll()) {
					testViewer.selectFirstFailure();
				}
			}
			return new Status(IStatus.OK, TestFrameworkPlugin.PLUGIN_ID, IStatus.OK,msg.getString("TestRunnerViewPart.OK"),null); //$NON-NLS-1$
		}
	}

	private class FailuresOnlyFilterAction extends Action {
		public FailuresOnlyFilterAction() {
			super(msg.getString("TestRunnerViewPart.ShowFailuresOnly"), AS_CHECK_BOX); //$NON-NLS-1$
			setToolTipText(msg.getString("TestRunnerViewPart.ShowFailuresOnly")); //$NON-NLS-1$
			setImageDescriptor(TestFrameworkPlugin.getImageDescriptor("obj16/failures.gif")); //$NON-NLS-1$
		}

		public void run() {
			setShowFailuresOnly(isChecked());
		}
	}
	
	private class RerunLastTestAction extends Action{
		public RerunLastTestAction() {
			setText(msg.getString("TestRunnerViewPart.RerunTest"));  //$NON-NLS-1$
			setToolTipText(msg.getString("TestRunnerViewPart.RerunTest")); //$NON-NLS-1$
			setDisabledImageDescriptor(TestFrameworkPlugin.getImageDescriptor("dlcl16/relaunch.gif")); //$NON-NLS-1$
			setHoverImageDescriptor(TestFrameworkPlugin.getImageDescriptor("obj16/relaunch.gif")); //$NON-NLS-1$
			setImageDescriptor(TestFrameworkPlugin.getImageDescriptor("obj16/relaunch.gif")); //$NON-NLS-1$
			setEnabled(false);
		}
		
		public void run(){
			rerunTestRun();
		}

	}
	
	private class StopAction extends Action{
		public StopAction() {
			setText(msg.getString("TestRunnerViewPart.StopCuteTestRun")); //$NON-NLS-1$
			setToolTipText(msg.getString("TestRunnerViewPart.StopCuteTestRun")); //$NON-NLS-1$
			setDisabledImageDescriptor(TestFrameworkPlugin.getImageDescriptor("dlcl16/stop.gif")); //$NON-NLS-1$
			setHoverImageDescriptor(TestFrameworkPlugin.getImageDescriptor("obj16/stop.gif")); //$NON-NLS-1$
			setImageDescriptor(TestFrameworkPlugin.getImageDescriptor("obj16/stop.gif")); //$NON-NLS-1$
		}

		public void run() {
			stopTest();
//			setEnabled(false);
		}
	}

	public void setShowFailuresOnly(boolean b) {
		testViewer.setFailuresOnly(b);		
	}

	public void stopTest() {
		if(session != null) {
			try {
				for(IProcess process : session.getLaunch().getProcesses()) {
					process.terminate();
				}
//				new SessionFinishedUIJob("Process Stopped").schedule();
			}catch(DebugException de) {
				
			}
		}
	}

	public void rerunTestRun() {
		if (session != null && session.getLaunch().getLaunchConfiguration() != null) {
			ILaunchConfiguration configuration= session.getLaunch().getLaunchConfiguration();
			DebugUITools.launch(configuration, session.getLaunch().getLaunchMode());
		}
		
	}

	public void selectNextFailure() {
		testViewer.selectNextFailure();		
	}
	
	public void selectPrevFailure() {
		testViewer.selectPrevFailure();
	}

	public void sessionFinished(TestSession session) {
		SessionFinishedUIJob sessionFinishedUIJob = new SessionFinishedUIJob(msg.getString("TestRunnerViewPart.SessionOver")); //$NON-NLS-1$
		sessionFinishedUIJob.schedule();
	}

	public void sessionStarted(TestSession session) {
		this.session = session;
		showNextFailureAction.setEnabled(false);
		showPreviousFailureAction.setEnabled(false);
		rerunLastTestAction.setEnabled(false);
		stopAction.setEnabled(true);
	}

}  //  @jve:decl-index=0:visual-constraint="148,36,771,201"

File: CUTEProject/bundles/ch.hsr.ifs.testframework/src/ch/hsr/ifs/testframework/ui/TestRunnerViewPart.kt
/*******************************************************************************
 * Copyright (c) 2007-2011, IFS Institute for Software, HSR Rapperswil,
 * Switzerland, http://ifs.hsr.ch
 *
 * Permission to use, copy, and/or distribute this software for any
 * purpose without fee is hereby granted, provided that the above
 * copyright notice and this permission notice appear in all copies.
 ******************************************************************************/
package ch.hsr.ifs.testframework.ui

import ch.hsr.ifs.testframework.TestFrameworkPlugin
import ch.hsr.ifs.testframework.model.ISessionListener
import ch.hsr.ifs.testframework.model.TestSession
import org.eclipse.cdt.debug.core.ICDTLaunchConfigurationConstants
import org.eclipse.core.runtime.CoreException
import org.eclipse.core.runtime.IProgressMonitor
import org.eclipse.core.runtime.IStatus
import org.eclipse.core.runtime.Status
import org.eclipse.debug.core.DebugException
import org.eclipse.debug.core.model.IProcess
import org.eclipse.debug.ui.DebugUITools
import org.eclipse.jface.action.Action
import org.eclipse.jface.action.IAction
import org.eclipse.jface.action.Separator
import org.eclipse.jface.viewers.TreeViewer
import org.eclipse.swt.SWT
import org.eclipse.swt.events.ControlEvent
import org.eclipse.swt.events.ControlListener
import org.eclipse.swt.layout.GridData
import org.eclipse.swt.layout.GridLayout
import org.eclipse.swt.widgets.Composite
import org.eclipse.ui.part.ViewPart
import org.eclipse.ui.progress.UIJob


public class TestRunnerViewPart : ViewPart(), ISessionListener {

	companion object {
		const val ID = "ch.hsr.ifs.cutelauncher.ui.TestRunnerViewPart"
		val msg = TestFrameworkPlugin.getMessages()
	}

	private enum class Orientation {
		horizontal, vertical
	}

	private lateinit var top: Composite
	private lateinit var TopPanel: Composite
	protected var autoScroll = true
	protected lateinit var counterPanel: CounterPanel
	private lateinit var cuteProgressBar: CuteProgressBar
	private lateinit var testViewer: TestViewer
	private lateinit var parent: Composite
	private var currentOrientation = Orientation.horizontal
	private lateinit var scrollLockAction: ScrollLockAction
	private lateinit var failureOnlyAction: FailuresOnlyFilterAction
	private lateinit var showNextFailureAction: ShowNextFailureAction
	private lateinit var showPreviousFailureAction: ShowPreviousFailureAction
	private var session: TestSession? = null
	private lateinit var stopAction: StopAction
	private lateinit var rerunSelectedAction: RerunSelectedAction

	init {
		TestFrameworkPlugin.getModel().addListener(this)
	}

	val rerunLastTestAction: IAction by lazy {
		RerunLastTestAction(this).apply {
			setEnabled(false)
		}
	}

	override fun createPartControl(parent: Composite) {
		this.parent = parent
		addResizeListener(parent)
		val gridLayout = GridLayout()
		gridLayout.numColumns = 1
		gridLayout.marginWidth = 0
		gridLayout.horizontalSpacing = 0
		gridLayout.marginWidth = 0
		gridLayout.marginHeight = 0
		val gdata = GridData()
		gdata.grabExcessHorizontalSpace = true
		top = Composite(parent, SWT.NONE)
		top.setLayout(gridLayout)
		top.setLayoutData(gdata)
		setPartName(msg.getString("TestRunnerViewPart.Name"))
		createTopPanel()
		createTestViewer()
		configureToolbar()
		getSite().setSelectionProvider(testViewer.getTreeViewer())
	}

	private fun addResizeListener(parent: Composite) {
		parent.addControlListener(object : ControlListener {

			override fun controlMoved(e: ControlEvent) = Unit

			override fun controlResized(e: ControlEvent) {
				computeOrientation()
			}
		})
	}

	val isCreated get() = this::counterPanel.isInitialized

	private fun configureToolbar() {
		val actionBars = getViewSite().getActionBars()
		val toolBar = actionBars.getToolBarManager()

		scrollLockAction = ScrollLockAction(this)
		scrollLockAction.setChecked(!autoScroll)

		failureOnlyAction = FailuresOnlyFilterAction()
		failureOnlyAction.setChecked(false)

		showNextFailureAction = ShowNextFailureAction(this)
		showNextFailureAction.setEnabled(false)
		showPreviousFailureAction = ShowPreviousFailureAction(this)
		showPreviousFailureAction.setEnabled(false)

		stopAction = StopAction()
		stopAction.setEnabled(false)

		toolBar.add(showNextFailureAction)
		toolBar.add(showPreviousFailureAction)
		toolBar.add(failureOnlyAction)
		toolBar.add(scrollLockAction)
		toolBar.add(Separator())
		toolBar.add(rerunLastTestAction)
		toolBar.add(stopAction)
	}

	private fun createTopPanel() {
		val gridLayout1 = GridLayout()
		gridLayout1.numColumns = 2
		val gridData = GridData()
		gridData.grabExcessHorizontalSpace = true
		gridData.verticalAlignment = org.eclipse.swt.layout.GridData.FILL
		gridData.horizontalAlignment = org.eclipse.swt.layout.GridData.FILL
		TopPanel = Composite(top, SWT.NONE)
		createCounterPanel()
		TopPanel.setLayout(gridLayout1)
		TopPanel.setLayoutData(gridData)
		createCuteProgressBar()
	}

	private fun createCounterPanel() {
		val gridData1 = GridData()
		gridData1.grabExcessHorizontalSpace = true
		gridData1.horizontalAlignment = org.eclipse.swt.layout.GridData.FILL
		counterPanel = CounterPanel(TopPanel, SWT.NONE)
		counterPanel.setLayoutData(gridData1)
	}

	private fun createCuteProgressBar() {
		val gridData2 = GridData()
		gridData2.grabExcessHorizontalSpace = true
		gridData2.horizontalIndent = 35
		gridData2.horizontalAlignment = org.eclipse.swt.layout.GridData.FILL
		cuteProgressBar = CuteProgressBar(TopPanel)
		cuteProgressBar.setLayoutData(gridData2)
	}

	private fun createTestViewer() {
		val gridData3 = GridData()
		gridData3.grabExcessHorizontalSpace = true
		gridData3.horizontalAlignment = org.eclipse.swt.layout.GridData.FILL
		gridData3.verticalAlignment = org.eclipse.swt.layout.GridData.FILL
		gridData3.grabExcessVerticalSpace = true
		testViewer = TestViewer(top, SWT.NONE, this)
		testViewer.setLayoutData(gridData3)
	}

	private fun computeOrientation() {
		val size = parent.getSize()
		if (size.x != 0 && size.y != 0) {
			if (size.x > size.y) setOrientation(Orientation.horizontal)
			else setOrientation(Orientation.vertical)
		}
	}

	private fun setOrientation(orientation: Orientation) {
		testViewer.setOrientation(orientation == Orientation.horizontal)
		currentOrientation = orientation
		val layout = TopPanel.getLayout() as GridLayout
		setCounterColumns(layout)
		parent.layout()
	}

	private fun setCounterColumns(layout: GridLayout) = when (currentOrientation) {
		Orientation.horizontal -> layout.numColumns = 2
		else -> layout.numColumns = 1
	}


	override fun setFocus() = Unit

	var isAutoScroll
		get() = autoScroll
		set(value) {
			autoScroll = value
		}

	private inner class SessionFinishedUIJob(name: String) : UIJob(name) {

		override fun runInUIThread(monitor: IProgressMonitor): IStatus {
			rerunLastTestAction.setEnabled(true)
			stopAction.setEnabled(false)
			if (this@TestRunnerViewPart.session?.hasErrorOrFailure() ?: false) {
				showNextFailureAction.setEnabled(true)
				showPreviousFailureAction.setEnabled(true)
				if (isAutoScroll) {
					testViewer.selectFirstFailure()
				}
			}
			return Status(IStatus.OK, TestFrameworkPlugin.PLUGIN_ID, IStatus.OK, msg.getString("TestRunnerViewPart.OK"), null)
		}
	}

	private inner class FailuresOnlyFilterAction : Action(msg.getString("TestRunnerViewPart.ShowFailuresOnly"), AS_CHECK_BOX) {

		init {
			setToolTipText(msg.getString("TestRunnerViewPart.ShowFailuresOnly"))
			setImageDescriptor(TestFrameworkPlugin.getImageDescriptor("obj16/failures.gif"))
		}

		override fun run() = setShowFailuresOnly(isChecked())
	}

	private inner class StopAction : Action() {

		init {
			setText(msg.getString("TestRunnerViewPart.StopCuteTestRun"))
			setToolTipText(msg.getString("TestRunnerViewPart.StopCuteTestRun"))
			setDisabledImageDescriptor(TestFrameworkPlugin.getImageDescriptor("dlcl16/stop.gif"))
			setHoverImageDescriptor(TestFrameworkPlugin.getImageDescriptor("obj16/stop.gif"))
			setImageDescriptor(TestFrameworkPlugin.getImageDescriptor("obj16/stop.gif"))
		}

		override fun run() = stopTest()
	}

	fun setShowFailuresOnly(b: Boolean) {
		testViewer.failuresOnly = b
	}

	fun stopTest(): Unit = try {
		session?.launch?.processes?.forEach(IProcess::terminate) ?: Unit
	} catch (e: DebugException) {
	}

	fun rerunTestRun() {
		session?.apply {
			launch?.launchConfiguration?.let {
				DebugUITools.launch(it, this.launch.launchMode)
			}
		}
	}

	fun rerunSelectedTestRun(rerunnames: List<String>) {
		session?.apply {
			launch?.launchConfiguration?.let {
				try {
					val copy = it.copy("")
					val builder = StringBuilder()
					rerunnames.forEach {
						builder.append(" \"$it\"")
					}
					copy.setAttribute(ICDTLaunchConfigurationConstants.ATTR_PROGRAM_ARGUMENTS, builder.toString())
					copy
				} catch (e: CoreException) {
					TestFrameworkPlugin.log(e)
					it
				}?.let {
					DebugUITools.launch(it, this.launch.launchMode)
				}
			}
		}
	}

	fun selectNextFailure() {
		testViewer.selectNextFailure()
	}

	fun selectPrevFailure() {
		testViewer.selectPrevFailure()
	}

	override fun sessionFinished(session: TestSession) {
		val sessionFinishedUIJob = SessionFinishedUIJob(msg.getString("TestRunnerViewPart.SessionOver"))
		sessionFinishedUIJob.schedule()
	}

	override fun sessionStarted(session: TestSession) {
		this.session = session
		showNextFailureAction.setEnabled(false)
		showPreviousFailureAction.setEnabled(false)
		rerunLastTestAction.setEnabled(false)
		stopAction.setEnabled(true)
	}

	fun getRerunSelectedAction(treeViewer: TreeViewer): RerunSelectedAction {
		if (!this::rerunSelectedAction.isInitialized) {
			rerunSelectedAction = RerunSelectedAction(this, treeViewer)
		}
		return rerunSelectedAction
	}

}
